### PR TITLE
Add available_nproc function

### PR DESCRIPTION
**Pull request by dagewa**
For #1408

### CODE REVIEW

**Comment by rjgildea** [comment]
Would it be possible to add some comments describing the circumstances under which each level of fallback will get hit? Presumably these are system/Python version dependent?

I guess testing this function is not trivial, although as a minimum it might be possible to test getting the value from the environment with `monkeypatch.setenv`:
https://docs.pytest.org/en/stable/monkeypatch.html

File: util/mp.py [hunk 1]
```diff
@@ -1,10 +1,51 @@
...
 import future.moves.itertools as itertools
+import psutil
 
 import libtbx.easy_mp
 
 
+def available_nproc():
+    """
+    Choose a suitable value for nproc based on the availability of processors.
+    """
```
Comment (Anthchirp) [inline util/mp.py:14]:
```suggestion
def available_cores() -> int:
    """
    Determine the number of available processor cores.
    There are a number of different methods to get this information, some of which may not
    be available on a specific OS and/or version of Python. So try them in order and return
    the first successful one.
    """
```

**Review by Anthchirp — APPROVED** [review]
I don't think comments are particularly helpful here, as they depend on particular versions of Python, psutils, and considerations within those packages...
Possibly more useful to put in links to the relevant documentations instead?

I would suggest a different function name though.

File: util/mp.py (hunk 1 above)
Comment (dagewa) [reply util/mp.py:14]: can do - took `nproc` as being the accepted jargon in DIALS, but `cores` seems better. Given your improved docstring I'll remove the comments I added

File: util/mp.py [hunk 2]
```diff
@@ -1,10 +1,51 @@
...
+
+    nproc = os.environ.get("NSLOTS", 0)
+    try:
+        nproc = int(nproc)
+    except ValueError:
+        nproc = 0
+    if nproc >= 1:
+        return nproc
```
Comment (Anthchirp) [inline util/mp.py:25]:
```suggestion
        nproc = int(nproc)
        if nproc >= 1:
            return nproc
    except ValueError:
        pass
```

Python 3 does not allow string / int comparison, so only consider returning the value in the `try` block

**Comment by dagewa** [comment]
When #1408 is fixed then this will get regular real world testing on the platforms we are interested in. Do we really want an additional unit test for this when the code is trivial?

**Comment by rjgildea** [comment]
> When #1408 is fixed then this will get regular real world testing on the platforms we are interested in. Do we really want an additional unit test for this when the code is trivial?

After #1408 is fixed, we should probably explicitly set `nproc` in most/all of the spotfinding tests in order to play nicely with parallel tests.

**Comment by phyy-nx** [comment]
Can pytests be marked to use N number of processes as part of pytest-xdist?  That would be helpful here.  We have been doing some of this in [libtbx land](https://github.com/cctbx/cctbx_project/commit/f5f88156b518a1acd6362e156cad679a93c9aebe) for OpenMP tests.  You can now mark some tests as parallel, meaning they are run in serial after all the other tests and use the number of processors specified in nproc.

**Comment by Anthchirp** [comment]
Just mock out the `available_cores` function? Or just use all cores. I don't believe this has a huge penalty?

**Comment by Anthchirp** [comment]
test timings: (`pytest test_mp.py  --durations=0 -vv`)
```
=== slowest test durations ===
0.00s setup    test/util/test_mp.py::test_core_detection
0.00s setup    test/util/test_mp.py::test_core_detection_override
0.00s teardown test/util/test_mp.py::test_core_detection_override
0.00s teardown test/util/test_mp.py::test_core_detection
0.00s call     test/util/test_mp.py::test_core_detection
0.00s call     test/util/test_mp.py::test_core_detection_override
```
we can afford that :slightly_smiling_face:

**Review by rjgildea — APPROVED** [review]
Looks good. Thanks for adding the test @Anthchirp 👍

**Pull request merged**